src/compiler/program.ts

```diff
@@ -1520,7 +1527,297 @@ namespace ts {
+                        if (activeLint.accepted) {
+                            oneAccepted = true;
+                        }
+                        else {
+                            needsReset[i] = true;
+                        }
+                    }
+                }
+                parent = node;
+                if (oneAccepted) {
+                    forEachChild(node, visitNode);
+                }
+                parent = oldParent;
+                for (let i = 0; i < initializedLints.length; i++) {
+                    activeLint = initializedLints[i];
+                    if (!needsReset[i]) {
+                        activeLint.accepted = true;
+                        needsReset[i] = false;
+                    }
+                    if (!activeLint.errored && activeLint.walker.afterVisit) {
+                        try {
+                            activeLint.walker.afterVisit(node, error);
+                        }
+                        catch (e) {
+                            activeLint.errored = true;
+                            diagnostics.push(createExtensionDiagnostic(errorQualifiedName("!!!"), `afterVisit failed with error: ${e}`, /*sourceFile*/undefined, /*start*/undefined, /*length*/undefined, DiagnosticCategory.Error));
+                        }
+                    }
+                }
+            }
+
+            function errorQualifiedName(shortname?: string) {
+                if (!shortname) return activeLint.name;
+                return `${activeLint.name}(${shortname})`;
+            }
+
+            function error(err: string): void;
+            function error(err: string, node: Node): void;
+            function error(err: string, start: number, length: number): void;
+            function error(shortname: string, err: string): void;
+            function error(shortname: string, err: string, span: Node): void;
+            function error(shortname: string, err: string, start: number, length: number): void;
+            function error(level: DiagnosticCategory, err: string): void;
+            function error(level: DiagnosticCategory, err: string, node: Node): void;
+            function error(level: DiagnosticCategory, err: string, start: number, length: number): void;
+            function error(level: DiagnosticCategory, shortname: string, err: string): void;
+            function error(level: DiagnosticCategory, shortname: string, err: string, span: Node): void;
+            function error(level: DiagnosticCategory, shortname: string, err: string, start: number, length: number): void;
+            function error(
+                levelErrOrShortname: DiagnosticCategory | string,
+                errShortnameSpanOrStart?: string | Node | number,
+                errSpanStartOrLength?: string | Node | number,
+                startSpanOrLength?: number | Node,
+                length?: number): void {
+                if (typeof levelErrOrShortname === "undefined") {

```

Comment:
Honestly? I don't think I've actually seen a JS function this polymorphic in a long time. Still, feels good knowing that every overload is uniquely distinguishable - I probably wouldn't have even tried to write something like this without TS to back me up on its validity.

Rather than checking each argument for undefined-ness, I could just check arguments.length, and dispatch on that. If I put the implementations into a hash, then I could simple have `return (errorOverloads[arguments.length] || tooManyArgs).call(undefined, arguments)`. If you think that'd look better/be more clear.